generator/src/lib.rs

```diff
@@ -2248,26 +2246,32 @@ fn generate_union(union: &vkxml::Union) -> TokenStream {
         }
     }
 }
-pub fn root_struct_names(definitions: &[&vkxml::DefinitionsElement]) -> HashSet<String> {
-    definitions
-        .iter()
-        .filter_map(|definition| match *definition {
-            vkxml::DefinitionsElement::Struct(ref _struct) => {
-                let is_root_struct = _struct.extends.is_none();
-                if is_root_struct {
-                    Some(_struct.name.clone())
-                } else {
-                    None
-                }
-            }
-            _ => None,
-        })
-        .collect()
+/// root structs are all structs that are extended by other structs
+pub fn root_structs(definitions: &[&vkxml::DefinitionsElement]) -> HashSet<Ident> {
+    let mut root_structs = HashSet::new();
+    // loop over all structs and collect their extends
+    for definition in definitions {
+        if let vkxml::DefinitionsElement::Struct(ref _struct) = definition {
+            let extends_structs = _struct
+                .extends
+                .as_ref()
+                .map(|extends| {
+                    extends
+                        .split(',')
+                        .map(|extends| name_to_tokens(extends))
+                        .collect()
+                })
+                .unwrap_or_else(HashSet::new);
+            root_structs.extend(extends_structs);
```

Comment:
`fn extend` takes an iterator so there's no need to collect this into an intermediary `HashSet` that is later only iterated. I won't be opposed to something like:

```suggestion
            if let Some(extends) = &_struct.extends {
                root_structs.extend(extends.split(',').map(name_to_tokens));
            }
```

Reply:
I did not like this code too. Thanks for the much better suggestion !

Same for this bit:
```rust
    let extends_structs = _struct
        .extends
        .as_ref()
        .map(|extends| {
            extends
                .split(',')
                .map(|extends| format_ident!("Extends{}", name_to_tokens(extends)))
                .collect()
        })
        .unwrap_or_else(Vec::new);
    let impl_extend_trait = extends_structs.iter().map(|extends| {
        quote! {
            unsafe impl #extends for #name_builder<'_> {}
            unsafe impl #extends for #name {}
        }
    });
```
Any suggestions there ?

Reply:
I had been looking at that one as well, unfortunately there's not much more we can do than:

```rust
    let impl_extend_trait = _struct
        .extends
        .as_ref()
        .map(|extends| {
            extends
                .split(',')
                .map(|extends| format_ident!("Extends{}", name_to_tokens(extends)))
                .map(|extends| {
                    quote! {
                        unsafe impl #extends for #name_builder<'_> {}
                        unsafe impl #extends for #name {}
                    }
                })
                .collect()
        })
        .unwrap_or_else(Vec::new);
```
Which might be slightly more memory-heavy as it has to keep the full tokentrees in memory instead of just the identifier, and expanding that when the final quote macro runs.

Unfortunately quote is able to expand `Option<>` and iterables, but not both at the same time - so the above without `collect()` and `unwrap_or_else` is not possible. Likewise `unwrap_or_else(std::iter::empty)` doesn't work either, because that's an iterator of a different type.

Reply:
So should I go for the proposed alternative or leave it as is ?

Reply:
Probably not the answer you were looking for, but I really don't have a preference :)

However, you might also move the `.map(|extends| format_ident!("Extends{}", name_to_tokens(extends)))` over to the iterator :D

Reply:
Done as suggested

Reply:
There is also `flat_map` which you could use on the option. `option.iter().flat_map`

Reply:
> 
> 
> There is also `flat_map` which you could use on the option. `option.iter().flat_map`

Done

Reply:
Ah `flatten` on `Option` is one of the few library bits I keep forgetting about :facepalm: 

Now you don't need the intermediary vector either:


```rust
    let impl_extend_trait = _struct
        .extends
        .iter()
        .flat_map(|extends| extends.split(','))
        .map(|extends| format_ident!("Extends{}", name_to_tokens(extends)))
        .map(|extends| {
            quote! {
                unsafe impl #extends for #name_builder<'_> {}
                unsafe impl #extends for #name {}
            }
        });
```